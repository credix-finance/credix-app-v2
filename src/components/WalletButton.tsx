import React, { useCallback, useMemo, useEffect } from "react";
import { useWalletModal, WalletIcon } from "@solana/wallet-adapter-react-ui";
import { useWallet } from "@solana/wallet-adapter-react";
import { useStore } from "@state/useStore";
import { Button } from "@components/Button";
import { Icon, IconDimension } from "@components/Icon";
import { useIntl } from "react-intl";
import { Menu, Dropdown } from "antd";
import notification from "@notification";

interface WalletButtonProps {
	className?: string;
}

export const WalletButton = ({ className = "" }: WalletButtonProps) => {
	const { wallet, publicKey, disconnect } = useWallet();
	const { setVisible } = useWalletModal();
	const base58 = useMemo(() => publicKey?.toBase58(), [publicKey]);
	const setIsAdmin = useStore((state) => state.setIsAdmin);
	const intl = useIntl();

	const address = useMemo(() => {
		if (!wallet || !base58) {
			return intl.formatMessage({
				defaultMessage: "Connecting",
				description: "WalletButton: connecting",
			});
		}

		return base58.slice(0, 5) + ".." + base58.slice(-5);
	}, [wallet, base58, intl]);

	const copyAddress = useCallback(async () => {
		if (base58) {
			try {
				await navigator.clipboard.writeText(base58);
				notification.success({
					message: intl.formatMessage({
						defaultMessage: "Copied to clipboard!",
						description: "WalletButton: copy address success",
					}),
				});
			} catch (error) {
				notification.error({
					message: intl.formatMessage({
						defaultMessage: "Failed to copy address to clipboard",
						description: "WalletButton: copy address failed",
					}),
					error,
				});
			}
		}
	}, [base58, intl]);

	const disconnectWallet = async () => {
		try {
			await disconnect();
			notification.success({
				message: intl.formatMessage({
					defaultMessage: "Wallet disconnected",
					description: "WalletButton: disconnect wallet success",
				}),
			});
		} catch (error) {
			notification.error({
				message: intl.formatMessage({
					defaultMessage: "Failed to disconnect wallet",
					description: "WalletButton: disconnect wallet failed",
				}),
				error,
			});
		}
	};

	useEffect(() => setIsAdmin(publicKey), [setIsAdmin, publicKey]);

	if (!wallet && !publicKey) {
		return (
			<Button
				onClick={() => setVisible(true)}
				icon={<Icon name="wallet" size={IconDimension.MIDDLE} />}
				className={className}
				data-cy="wallet-button"
			>
				<span className="text-lg font-semibold capitalize">
					{intl.formatMessage({
						defaultMessage: "connect wallet",
						description: "WalletButton: connect button",
					})}
				</span>
			</Button>
		);
	}

	const menu = (
		<Menu className="bg-credix-primary shadow-none border-solid border-[1px]">
			<Menu.Item className="bg-credix-primary hover:bg-neutral-10" onClick={copyAddress}>
				{intl.formatMessage({
					defaultMessage: "copy address",
					description: "WalletButton: copy address button",
				})}
			</Menu.Item>
			<Menu.Item className="bg-credix-primary hover:bg-neutral-10" onClick={() => setVisible(true)}>
				{intl.formatMessage({
					defaultMessage: "change wallet",
					description: "WalletButton: change wallet button",
				})}
			</Menu.Item>
			<Menu.Item
				data-cy="disconnect-wallet-button"
				className="bg-credix-primary hover:bg-neutral-10"
				onClick={disconnectWallet}
			>
				{intl.formatMessage({
					defaultMessage: "disconnect",
					description: "WalletButton: disconnect button",
				})}
			</Menu.Item>
		</Menu>
	);

	return (
		<Dropdown overlay={menu} trigger={["click"]}>
			<Button
				type="default"
				onClick={(event) => event.preventDefault()}
				icon={<WalletIcon wallet={wallet} className="w-6" />}
				className={`${className} w-56`}
			>
				<span data-cy="wallet-address">{address}</span>
			</Button>
		</Dropdown>
	);
};
