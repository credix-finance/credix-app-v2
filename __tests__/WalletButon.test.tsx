import React from "react";
import renderer from "react-test-renderer";
import { WalletButton } from "@components/WalletButton";

test("Wallet button", () => {
	const component = renderer.create(<WalletButton />);

	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
