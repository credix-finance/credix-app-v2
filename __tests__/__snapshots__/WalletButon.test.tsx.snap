// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet button 1`] = `
<button
  className="ant-btn ant-btn-primary ant-btn-lg rounded-[1px] border text-shadow-none shadow-none flex items-center justify-center gap-2 px-[25px] 
				bg-action-primary border-action-primary text-credix-primary
				hover:bg-neutral-60 hover:border-neutral-60
				active:bg-neutral-40 active:border-neutral-40 active:text-neutral-100
				disabled:bg-action-disable disabled:border-transparent disabled:text-disabled
			 text-lg font-semibold h-[50px] "
  onClick={[Function]}
  type="button"
>
  <svg
    className="w-4 h-4 fill-current w-6 h-6"
    viewBox="0 0 24 25"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M17 9.5H22V15.5H17C15.3431 15.5 14 14.1569 14 12.5C14 10.8431 15.3431 9.5 17 9.5ZM17 13.5C17.5523 13.5 18 13.0523 18 12.5C18 11.9477 17.5523 11.5 17 11.5C16.4477 11.5 16 11.9477 16 12.5C16 13.0523 16.4477 13.5 17 13.5Z"
      fillRule="evenodd"
    />
    <path
      clipRule="evenodd"
      d="M7.7587 4.5H16.2413C17.0463 4.49999 17.7106 4.49998 18.2518 4.54419C18.8139 4.59012 19.3306 4.68868 19.816 4.93597C20.5686 5.31947 21.1805 5.93139 21.564 6.68404C21.8113 7.16937 21.9099 7.68608 21.9558 8.24817C22 8.78936 22 9.45372 22 10.2587V14.7413C22 15.5463 22 16.2106 21.9558 16.7518C21.9099 17.3139 21.8113 17.8306 21.564 18.316C21.1805 19.0686 20.5686 19.6805 19.816 20.064C19.3306 20.3113 18.8139 20.4099 18.2518 20.4558C17.7106 20.5 17.0463 20.5 16.2413 20.5H7.75872C6.95374 20.5 6.28938 20.5 5.74817 20.4558C5.18608 20.4099 4.66937 20.3113 4.18404 20.064C3.43139 19.6805 2.81947 19.0686 2.43597 18.316C2.18868 17.8306 2.09012 17.3139 2.04419 16.7518C1.99998 16.2106 1.99999 15.5463 2 14.7413V10.2587C1.99999 9.45373 1.99998 8.78937 2.04419 8.24818C2.09012 7.68608 2.18868 7.16937 2.43597 6.68404C2.81947 5.93139 3.43139 5.31947 4.18404 4.93597C4.66937 4.68868 5.18608 4.59012 5.74817 4.54419C6.28937 4.49998 6.95373 4.49999 7.7587 4.5ZM5.91104 6.53755C5.47262 6.57337 5.24842 6.6383 5.09202 6.71799C4.7157 6.90973 4.40973 7.2157 4.21799 7.59202C4.1383 7.74842 4.07337 7.97262 4.03755 8.41104C4.00078 8.86113 4 9.44342 4 10.3V14.7C4 15.5566 4.00078 16.1389 4.03755 16.589C4.07337 17.0274 4.1383 17.2516 4.21799 17.408C4.40973 17.7843 4.7157 18.0903 5.09202 18.282C5.24842 18.3617 5.47262 18.4266 5.91104 18.4624C6.36113 18.4992 6.94342 18.5 7.8 18.5H16.2C17.0566 18.5 17.6389 18.4992 18.089 18.4624C18.5274 18.4266 18.7516 18.3617 18.908 18.282C19.2843 18.0903 19.5903 17.7843 19.782 17.408C19.8617 17.2516 19.9266 17.0274 19.9624 16.589C19.9992 16.1389 20 15.5566 20 14.7V10.3C20 9.44342 19.9992 8.86113 19.9624 8.41104C19.9266 7.97262 19.8617 7.74842 19.782 7.59202C19.5903 7.2157 19.2843 6.90973 18.908 6.71799C18.7516 6.6383 18.5274 6.57337 18.089 6.53755C17.6389 6.50078 17.0566 6.5 16.2 6.5H7.8C6.94342 6.5 6.36113 6.50078 5.91104 6.53755Z"
      fillRule="evenodd"
    />
  </svg>
  <span
    className="text-lg font-semibold"
  >
    Connect Wallet
  </span>
</button>
`;
